google/cloud/pubsub/internal/tracing_batch_callback.cc

```diff
@@ -137,16 +139,51 @@ class TracingBatchCallback : public BatchCallback {
   }
 
  private:
+  struct MessageSpans {
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> subscribe_span;
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>
+        concurrency_control_span;
+  };
+
+  void StartConcurrencyControl(std::string const& ack_id) override {
+    namespace sc = opentelemetry::trace::SemanticConventions;
+    std::lock_guard<std::mutex> lk(mu_);
+    auto spans = spans_by_ack_id_.find(ack_id);
+    if (spans != spans_by_ack_id_.end()) {
+      auto subscribe_span = spans->second.subscribe_span;
+      if (subscribe_span) {
+        opentelemetry::trace::StartSpanOptions options;
+        options.parent = subscribe_span->GetContext();
+        auto span =
+            internal::MakeSpan("subscriber concurrency control",
+                               {{sc::kMessagingSystem, "gcp_pubsub"}}, options);
+        auto scope = internal::OTelScope(span);
+        spans->second.concurrency_control_span = span;
+      }
+    }
+  }
+
+  void EndConcurrencyControl(std::string const& ack_id) override {
+    std::lock_guard<std::mutex> lk(mu_);
+    auto spans = spans_by_ack_id_.find(ack_id);
+    if (spans != spans_by_ack_id_.end()) {
+      auto concurrency_control_span = spans->second.concurrency_control_span;
```

Comment:
Do you want to move `spans->second.concurrency_control_span` so that it is null when this unlocks?

Reply:
Yes